src/org/exist/collections/MutableCollection.java

```diff
@@ -1258,6 +1258,9 @@ private void storeXMLInternal(final Txn transaction, final DBBroker broker, fina
         if(info.isCreating()) {
             info.getTriggers().afterCreateDocument(broker, transaction, document);
         } else {
+            final StreamListener listener = broker.getIndexController().getStreamListener(document, StreamListener.ReindexMode.REPLACE_DOCUMENT);
```

Comment:
the difference that you do
```
setMode(REPLACE_DOCUMENT)
startReplaceDocument()

setMode(REMOVE_ALL_NODES)
startIndexDocument()
...
endIndexDocument()

setMode(STORE)
startIndexDocument()
...
endIndexDocument()

**setMode(REPLACE_DOCUMENT)**
endReplaceDocument()```

That mean `startIndexDocument` + `endIndexDocument` happen for same listener on `REMOVE_ALL_NODES` & `STORE`, but your `startReplaceDocument` + `endReplaceDocument` happen on different listeners. That is design bug, IMHO.